ykrt/src/compile/jitc_yk/codegen/mod.rs

```diff
@@ -0,0 +1,218 @@
+//! The JIT's Code Generator.
+
+// FIXME: eventually delete.
+#![allow(dead_code)]
+
+use super::{jit_ir, CompilationError};
+use dynasmrt::ExecutableBuffer;
+#[cfg(any(debug_assertions, test))]
+use dynasmrt::{AssemblyOffset, DynasmApi};
+#[cfg(any(debug_assertions, test))]
+use std::{cell::Cell, collections::HashMap, slice};
+use typed_index_collections::TiVec;
+
+/// FIXME: one arch is hard-compiled-in, but we should be able to "cross-unit-test" different
+/// architectures.
+#[cfg(target_arch = "x86_64")]
+mod x86_64;
+
+/// Describes storage for a local variable.
+///
+/// FIXME: This assumes everything is a stack slot. When we come to implement a register allocator,
+/// this should be an enum of stack/reg.
+#[derive(Debug, Clone, Copy)]
+struct LocalAlloc {
+    /// The offset from the base pointer of the allocation.
+    ///
+    /// This is independent of which direction the stack grows. In other words, for architectures
+    /// where the stack grows downwards, you'd subtract this from the base pointer to find the
+    /// address of the allocation.
+    ///
+    /// FIXME: when we have register allocation, consider addressing stack slots by the stack
+    /// pointer, thus freeing up the base pointer for general purpose use.
+    frame_off: usize,
+}
+
+impl LocalAlloc {
+    fn new(frame_off: usize) -> Self {
+        Self { frame_off }
+    }
+}
+
+impl LocalAlloc {
+    fn frame_off(&self) -> usize {
+        self.frame_off
+    }
+}
+
+/// The result of running the code generator.
+pub(crate) struct CodeGenOutput {
+    /// The executable code itself.
+    buf: ExecutableBuffer,
+    /// Comments to be shown when printing the compiled trace using `AsmPrinter`.
+    ///
+    /// Used for testing and debugging.
+    #[cfg(any(debug_assertions, test))]
+    comments: HashMap<AssemblyOffset, Vec<String>>,
+}
+
+impl CodeGenOutput {
+    #[cfg(any(debug_assertions, test))]
+    fn disassemble(&self) -> String {
+        let tp = AsmPrinter::new(&self.buf);
+        #[cfg(any(debug_assertions, test))]
+        let tp = tp.set_comments(&self.comments);
+        tp.to_string()
+    }
+}
+
+/// The code generator.
+pub(crate) struct CodeGen<'a> {
+    jit_mod: &'a jit_ir::Module,
+    #[cfg(target_arch = "x86_64")]
+    asm: dynasmrt::x64::Assembler,
```

Comment:
I think we're immediately baking in `x86_64` here in such a way that we won't unbake it later. So either we call this `X86CodeGen` and don't pretend it's anything else, or we split things into machine independent and machine dependent parts now and get it (as) right (as we can).

Reply:
There's a comment to that effect someplace, and this is what I meant with "cross-arch testing" in the PR description.

As I said, I don't want to tackle this in the first PR, but if you think it's worthwhile (and I do think it is), we should address this in a follow up PR.

Reply:
Fixed in https://github.com/ykjit/yk/pull/971/commits/4c25a723383a8c577fad6d1f339b8dd37bd0c88e